In [1]:
# Import Libraries
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

from xgboost import XGBRegressor

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

In [2]:
# Load Dataset
df = pd.read_csv("../Dataset.csv")

df = df.drop_duplicates()

X = df.drop("Strength", axis=1)
y = df["Strength"]

In [3]:
# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

In [4]:
# Create Base Model
xgb = XGBRegressor(
    random_state=42
)

# Define Parameter Grid
param_grid = {

    'n_estimators': [100, 200],

    'max_depth': [4, 6, 8],

    'learning_rate': [0.01, 0.05, 0.1],

    'subsample': [0.8, 1.0],

    'colsample_bytree': [0.8, 1.0]

}

In [5]:
# Perform Grid Search
grid_search = GridSearchCV(

    estimator=xgb,

    param_grid=param_grid,

    cv=5,

    scoring='r2',

    n_jobs=-1,

    verbose=1

)

In [6]:
# Train Grid Search
grid_search.fit(X_train, y_train)

# Best Parameters
print("Best Parameters:")
print(grid_search.best_params_)

Fitting 5 folds for each of 72 candidates, totalling 360 fits
Best Parameters:
{'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 200, 'subsample': 0.8}


In [7]:
# Best Model
best_xgb = grid_search.best_estimator_

#  Prediction
prediction = best_xgb.predict(X_test)

In [8]:
# Evaluation
mae = mean_absolute_error(y_test, prediction)

mse = mean_squared_error(y_test, prediction)

rmse = np.sqrt(mse)

r2 = r2_score(y_test, prediction)

print("Optimized XGBoost")

print("MAE :", mae)

print("MSE :", mse)

print("RMSE :", rmse)

print("R2 Score :", r2)

Optimized XGBoost
MAE : 3.0207448262361742
MSE : 18.336404426739566
RMSE : 4.282102804317006
R2 Score : 0.9385354522400305


In [9]:
# Compare Before and After Tuning

comparison = pd.DataFrame({

    "Model":[
        "Default XGBoost",
        "Optimized XGBoost"
    ],

    "MAE":[
        2.9719,
        mae
    ],

    "RMSE":[
        4.5615,
        rmse
    ],

    "R2 Score":[
        0.9303,
        r2
    ]

})

comparison

,Model,MAE,RMSE,R2 Score
0,Default XGBoost,2.971900,4.561500,0.930300
1,Optimized XGBoost,3.020745,4.282103,0.938535


In [10]:
# Save the Best Model

import joblib

joblib.dump(best_xgb, "../saved_models/best_xgboost_model.pkl")

print("Model saved successfully!")

Model saved successfully!


## Observations

- Hyperparameter tuning was successfully performed on the **XGBoost Regressor** using **GridSearchCV** with **5-fold cross-validation**.

- A total of **72 hyperparameter combinations** (360 model fits) were evaluated to identify the optimal configuration for the dataset.

- The best hyperparameters obtained were:
  - **n_estimators = 200**
  - **max_depth = 4**
  - **learning_rate = 0.1**
  - **subsample = 0.8**
  - **colsample_bytree = 0.8**

- The optimized XGBoost model achieved an **MAE of 3.0207**, **MSE of 18.3364**, **RMSE of 4.2821**, and an **R² Score of 0.9385**.

- Compared to the default XGBoost model (**R² = 0.9303**), the optimized model improved the prediction performance by achieving a higher **R² Score of 0.9385** and a lower **RMSE**, indicating better generalization on the test dataset.

- Based on the evaluation results, the **Optimized XGBoost model** is selected as the **final prediction model** for the proposed Explainable Concrete Mix Optimization System.

- The optimized model will be used in the subsequent stages of the project, including **SHAP Explainability**, **Cost Estimation**, **Carbon Emission Analysis**, **Sustainability Assessment**, and **Concrete Mix Recommendation**.